test/test_dataloader.py

```diff
@@ -231,30 +231,50 @@ def __len__(self):
         return self.size
 
 
+class WorkerSpecificIterableDataset(IterableDataset):
+    def __init__(self, sizes_for_all_workers):
+        self.sizes_for_all_workers = sizes_for_all_workers
+
+    def __iter__(self):
```

Comment:
@apaszke Basically you can config each dataset replica differently in two different ways using `torch.utils.data.get_worker_info()` which returns the worker's `id`, `seed` and `dataset` replica:
1. In `__iter__` for an iterable dataset. 
2. In `worker_init_fn`.

This requires user to write their dataset code with multiprocessing data loading in mind. But I think this is a reasonable requirement because (1) there is no general way to split an iterator across multiple processes (2) if people want to do sharding / bulk loading, how to split work among workers should be part of the design.